# Name: Ejegu Smith
# Class: DTSC 680
# Assignment: 6

In [1]:
#common imports
import pandas as pd 
import numpy as np
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier

1. Import the Mushroom data set

In [2]:
col_names = ['class','cap-shape', 'cap-surface', 'cap-color', 'bruises?', 'odor', 'gill-attachment', 'gill-spacing',
            'gill-size', 'gill-color', 'stalk-shape', 'stalk-root', 'stalk-surface-above-ring', 
             'stalk-surface-below-ring', 'stalk-color-above-ring',  'stalk-color-below-ring',
            'veil-type', 'veil-color', 'ring-number', 'ring-type', 'spore-print-color',
            'population', 'habitat']

df = pd.read_csv('agaricus-lepiota.data', names=col_names)
df.shape

(8124, 23)

2. Use the KNN algorithm to impute missing values in the dataset. Note: You are not
permitted to use the KNNImputer class from Scikit-learn. Instead, you must
explicitly write the code to perform the imputation using the KNeighborsClassifier
algorithm (use the default settings).

In [4]:
#setting features and response for the full data set. Need to predict stalk-root here since this column has the missing data
response_all = df[['stalk-root']]
features_all = df.drop(['stalk-root'], axis=1)

In [5]:
# now one hot encoding our features data and one label encoding the response data
lab_all = LabelEncoder()
response_all = lab_all.fit_transform(response_all)

one_hot_all = OneHotEncoder(sparse = False)
features_all = one_hot_all.fit_transform(features_all)

/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [6]:
#checking 
lab_all.classes_

array(['?', 'b', 'c', 'e', 'r'], dtype=object)

In [7]:
#checking
np.unique(response_all)

array([0, 1, 2, 3, 4])

In [14]:
#changing our label-encoded response column to a data frame for further manipulation
df_response_all = pd.DataFrame(response_all)
df_response_all.columns = ['stalk-root']
df_response_all

stalk-root
0              3
1              2
2              2
3              3
4              3
...          ...
8119           0
8120           0
8121           0
8122           0
8123           0

[8124 rows x 1 columns]

In [16]:
#changing features to a data frame
df_features_all = pd.DataFrame(features_all)
df_features_all.head()

0    1    2    3    4    5    6    7    8    9    ...  104  105  106  107  \
0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  1.0  0.0  0.0  0.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
2  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  1.0  0.0  0.0  0.0   
4  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   

   108  109  110  111  112  113  
0  0.0  0.0  0.0  0.0  1.0  0.0  
1  1.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  1.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  1.0  0.0  
4  1.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 114 columns]

In [17]:
#now combining features and response encoded data frames
df_fullEncoded  = pd.concat([df_response_all, df_features_all], axis=1)
df_fullEncoded.shape

(8124, 115)

In [104]:
#checking
df_fullEncoded

stalk-root    0    1    2    3    4    5    6    7    8  ...  104  105  \
0              3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  1.0  0.0   
1              2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
2              2  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
3              3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  1.0  0.0   
4              3  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
...          ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
8119           0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0   
8120           0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0   
8121           0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8122           0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  1.0   
8123           0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   

      106  107  108  109  110  111  112  113  
0     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
1     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
2     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
3     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
4     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
...   ...  ...  ...  ...  ...  ...  ...  ...  
8119  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8120  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8121  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8122  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8123  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  

[8124 rows x 115 columns]

In [22]:
# isolating dataframe with missing values. 
# note 0 is label encoded for '?'
df_missing = df_fullEncoded[df_fullEncoded['stalk-root'] == 0]
df_missing

stalk-root    0    1    2    3    4    5    6    7    8  ...  104  105  \
3984           0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
4023           0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0   
4076           0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
4100           0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0   
4104           0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0   
...          ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
8119           0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0   
8120           0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0   
8121           0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8122           0  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  1.0   
8123           0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   

      106  107  108  109  110  111  112  113  
3984  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
4023  0.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  
4076  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  
4100  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  
4104  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
...   ...  ...  ...  ...  ...  ...  ...  ...  
8119  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8120  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8121  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8122  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
8123  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  

[2480 rows x 115 columns]

In [23]:
#now need to reomove missing values from the fully encoded dataset. we will call this new df, df_noMissing
df_noMissing = df_fullEncoded[df_fullEncoded['stalk-root'] != 0]
df_noMissing

stalk-root    0    1    2    3    4    5    6    7    8  ...  104  105  \
0              3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  1.0  0.0   
1              2  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
2              2  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
3              3  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  1.0  0.0   
4              3  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
...          ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
7986           1  1.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   
8001           1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
8038           1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
8095           2  0.0  1.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0   
8114           2  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0   

      106  107  108  109  110  111  112  113  
0     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
1     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
2     0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  
3     0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  
4     0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
...   ...  ...  ...  ...  ...  ...  ...  ...  
7986  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  
8001  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  
8038  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  
8095  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  
8114  0.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5644 rows x 115 columns]

 - Next, you should train your KNN model on
those instances that don’t have missing values, then have the model make predictions
for those instances that have missing values. This is how you will have the KNN model
impute missing data.

In [28]:
#need to create a features and response variable using df_noMissing (created in previous step) in order to train the model with no missing data
features_noMissing = df_noMissing.drop(['stalk-root'], axis=1)
response_noMissing = df_noMissing[['stalk-root']]

In [29]:
features_noMissing.shape

(5644, 114)

In [27]:
response_noMissing.shape

(5644, 1)

In [30]:
# now training the model
knn_clf1 = KNeighborsClassifier(n_neighbors=5)

knn_clf1.fit(features_noMissing, response_noMissing)

<ipython-input-30-eb89cdfc77af>:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  knn_clf1.fit(features_noMissing, response_noMissing)


KNeighborsClassifier()

In [32]:
X_test = df_fullEncoded.drop(['stalk-root']==0, axis=1)
X_test.shape

(8124, 114)

In [33]:
#predicting missing values for stalk-root
stalk_root_preds = knn_clf1.predict(X_test)
stalk_root_preds.shape

(8124,)

In [35]:
#changing back to categorical variables as required by the assignment
preds = lab_all.inverse_transform(stalk_root_preds) 

In [39]:
#turing preds into a df
preds_df = pd.DataFrame(preds)
preds_df

0
0     e
1     c
2     c
3     e
4     e
...  ..
8119  e
8120  e
8121  e
8122  b
8123  e

[8124 rows x 1 columns]

In [43]:
Actual = np.array(df_response_all).reshape(8124,)
values = pd.DataFrame({'Actual': Actual, 'Predicted': preds})
values

Actual Predicted
0          3         e
1          2         c
2          2         c
3          3         e
4          3         e
...      ...       ...
8119       0         e
8120       0         e
8121       0         e
8122       0         b
8123       0         e

[8124 rows x 2 columns]

- When you have computed the missing values, create a data structure (i.e. a list) called
missing_values that contains all of the imputed values (in terms of the original
categorical/letter data and not the encoded numeric data) in order of increasing index
from the original data set. You must then print the first 10 instances of
missing_values to the screen so we can check your work

In [85]:
missing_values = values[values.Actual == 0]
missing_values[['Predicted']]

Predicted
3984         b
4023         b
4076         e
4100         b
4104         b
...        ...
8119         e
8120         e
8121         e
8122         b
8123         e

[2480 rows x 1 columns]

- Finally, you will impute the
missing values back into the original data set before continuing, so that the next step
starts fresh with a complete data set in terms of the raw data values

In [48]:
#replacing missing values with predictions in original data set
df['stalk-root'] = preds_df

In [49]:
#checking to make sure there are no missing values in stalk-root column
df['stalk-root'].unique()

array(['e', 'c', 'b', 'r'], dtype=object)

In [51]:
#new data set with predictions injected
df

class cap-shape cap-surface cap-color bruises? odor gill-attachment  \
0        p         x           s         n        t    p               f   
1        e         x           s         y        t    a               f   
2        e         b           s         w        t    l               f   
3        p         x           y         w        t    p               f   
4        e         x           s         g        f    n               f   
...    ...       ...         ...       ...      ...  ...             ...   
8119     e         k           s         n        f    n               a   
8120     e         x           s         n        f    n               a   
8121     e         f           s         n        f    n               a   
8122     p         k           y         n        f    y               f   
8123     e         x           s         n        f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

## Graded Concept Question #1

Would it still be
possible to train the KNN model if you one-hot encoded the response data instead?
Why or why not?

Yes, it would still be possible as machine learning models need numerical inputs. In this case, the target data is either e for edible or p for poisionus. Therefore the target data must be converted to 1s and 0s. Using one-hot encoder on the target column would still result in the data being transformed from categorical to numerical, and thus the KNN model would be able to accept the target data.Furthermore, the target column is nominal instead of oridinal (no logical ordering), therefore one hot encoder can be used on the column. 

## Now making models to predict for class (edible or poisonous)...

In [52]:
response = df[['class']]
features = df.drop(['class'], axis=1)

In [53]:
#label encode response data
#1 = posionous
#0 = edible


lab_enc = LabelEncoder()
response = lab_enc.fit_transform(response)
response

/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


array([1, 0, 0, ..., 0, 1, 0])

In [54]:
#checking
lab_enc.classes_

array(['e', 'p'], dtype=object)

In [55]:
#checking
response.shape

(8124,)

In [56]:
#now one hot encode categorical features
ohe = OneHotEncoder(sparse = False)
features = ohe.fit_transform(features)
features

array([[0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [57]:
#now splitting test and training data
X_train, X_test, y_train, y_test = train_test_split(features, response, test_size=.20, random_state=42)

In [58]:
# creating and training the KNN model, then preforming predictions
knn_clf = KNeighborsClassifier()

knn_clf.fit(X_train, y_train)

preds = knn_clf.predict(X_test)

In [59]:
#KNN model accuracy score
from sklearn.metrics import accuracy_score

accuracy_score(preds, y_test)

1.0

3.Train a RandomForestClassifier as well as a LogisticRegression model to predict whether a mushroom is edible or poisonous given this data set of nominally-valued characteristics. Train the model on the feature data supplied to you after you’ve one-hot encoded it. You should label encode the response data.

In [60]:
from sklearn.ensemble import RandomForestClassifier

rnd_clf = RandomForestClassifier(random_state=42)

rnd_clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [61]:
y_pred_rnd_clf = rnd_clf.predict(X_test)
y_pred_rnd_clf

array([0, 1, 1, ..., 1, 1, 1])

In [62]:
#random forest classifier accuracy score
accuracy_score(y_pred_rnd_clf, y_test)

1.0

now logistic reg...

In [63]:
from sklearn.linear_model import LogisticRegression

log_clf = LogisticRegression(random_state=42)

log_clf.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [64]:
y_preds_log_clf = log_clf.predict(X_test)
y_preds_log_clf

array([0, 1, 1, ..., 1, 1, 1])

In [65]:
#Logistic Regression classifier accuracy score
accuracy_score(y_preds_log_clf, y_test)

1.0

## Graded Concept Question #2

Would it still be possible to train these two models if you one-hot encoded the response data instead,
being careful to specify that the drop parameter of the OneHotEncoder class is set to
‘first’? Why or why not?

No, it isn't possible. I recieved the following ValueError: "sparse multilabel-indicator for y is not supported." This is due to the curse of dimensionality problem. Additionally, the models do not support multiple columns for y, which is the result of applying the one hot encoder (ie one coulmn for edible and one column for posionous each with rows of 1's and 0's ). 

4. When you train both the RandomForestClassifier and LogisticRegression models, use the magic command %%time to time how long it takes to complete training. See the Python Data Science Handbook for more information about this magic command.

In [87]:
%%time
print('Logistic Regression time:')
log_clf.fit(X_train, y_train)

Logistic Regression time:
CPU times: user 263 ms, sys: 28.2 ms, total: 291 ms
Wall time: 256 ms


LogisticRegression(random_state=42)

In [89]:
%time
print('Random Forest Classifier time:')
rnd_clf.fit(X_train, y_train)
#19 nano-secs

CPU times: user 7 µs, sys: 12 µs, total: 19 µs
Wall time: 83.9 µs
Random Forest Classifier time:


RandomForestClassifier(random_state=42)

5. Compute the accuracy, precision, and recall scores for a test set. Briefly discuss the
performance of your models in terms of these values.

In [68]:
#random forest classifier scores
from sklearn.metrics import recall_score, precision_score

acc_score = accuracy_score(y_pred_rnd_clf, y_test)
rec_score = recall_score(y_test, y_pred_rnd_clf, average="micro")
prec_score = precision_score(y_test, y_pred_rnd_clf, average="micro")

print("Accuracy score:", round(acc_score,2))
print('Recall score:', round(rec_score,2))
print('Precision score:', round(prec_score,2))

Accuracy score: 1.0
Recall score: 1.0
Precision score: 1.0


In [69]:
#logistic regression scores

acc_score = accuracy_score(y_preds_log_clf, y_test)
rec_score = recall_score(y_test, y_preds_log_clf, average="micro")
prec_score = precision_score(y_test, y_preds_log_clf, average="micro")

print("Accuracy score:", round(acc_score,2))
print('Recall score:', round(rec_score,2))
print('Precision score:', round(prec_score,2))

Accuracy score: 1.0
Recall score: 1.0
Precision score: 1.0


 - Briefly discuss the
performance of your models in terms of these values.

All scores are 100%, which means the models are severly over-fitting. 

6. Perform dimensionality reduction using PCA and keep 95% of the variance. By what
percentage were you able to reduce the number of dimensions of the training set? How
many features (i.e. dimensions) are you left with after reducing dimensionality?

In [124]:
df_fullEncoded.shape

(8124, 115)

In [106]:
from sklearn.decomposition import PCA
import numpy as np

pca = PCA(n_components=0.95)
X_reduced = pca.fit_transform(df_fullEncoded.drop(columns='stalk-root', axis=1))

In [111]:
X_reduced.shape

(8124, 40)

In [108]:
#19% of the datasets variance lies on the 1st axis, and 12% on the 2nd axis, and so on
pca.explained_variance_ratio_

array([0.19373771, 0.12588053, 0.08966388, 0.06248633, 0.04038621,
       0.03724601, 0.03187816, 0.02692197, 0.02594401, 0.02105137,
       0.01986376, 0.01859088, 0.01810174, 0.01734219, 0.01546766,
       0.01485049, 0.01479827, 0.01408946, 0.01335092, 0.01163263,
       0.01101947, 0.01015416, 0.00981862, 0.00930722, 0.00868806,
       0.00818158, 0.00795653, 0.00743737, 0.00729614, 0.00691077,
       0.00649772, 0.00631657, 0.00578828, 0.00542183, 0.00529918,
       0.00488812, 0.00480136, 0.00471784, 0.00438651, 0.00425981])

In [112]:
X_train, X_test, y_train, y_test = train_test_split(X_reduced, stalk_root_preds, test_size=.20, random_state=42)

 - By what percentage were you able to reduce the number of dimensions of the training set? How many features (i.e. dimensions) are you left with after reducing dimensionality?

We went from 115 features to 40, this includes the expanded dimension data set due to one hot encoded columns. Overall this roughly a 65% reduction. 

In [127]:
#calcualtion
((40-115)/115)*100

-65.21739130434783

7. Train two new models, a RandomForestClassifier and a LogisticRegression model,
on this reduced dataset to predict whether a mushroom is edible or not. Again, time the
training of these two models on the reduced dataset. You will again use the default
hyperparameter values for both models, with a random state parameter of 42 for the
RandomForestClassifier.

In [122]:
%%time
rnd_clf_reduced = RandomForestClassifier(random_state=42)

rnd_clf_reduced.fit(X_train, y_train)

CPU times: user 2.37 s, sys: 19.3 ms, total: 2.39 s
Wall time: 2.42 s


RandomForestClassifier(random_state=42)

In [123]:
%%time

log_clf_reduced = LogisticRegression(random_state=42)

log_clf_reduced.fit(X_train, y_train)

CPU times: user 477 ms, sys: 36.2 ms, total: 513 ms
Wall time: 365 ms


/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [115]:
preds_reduced = log_clf_reduced.predict(X_test)

In [116]:
#logistic regression reduced scores

acc_score = accuracy_score(preds_reduced, y_test)
rec_score = recall_score(y_test, preds_reduced, average="micro")
prec_score = precision_score(y_test, preds_reduced, average="micro")

print("Accuracy score:", round(acc_score,2))
print('Recall score:', round(rec_score,2))
print('Precision score:', round(prec_score,2))

Accuracy score: 0.98
Recall score: 0.98
Precision score: 0.98


In [117]:
preds_red = rnd_clf_reduced.predict(X_test)

In [121]:
#random forest  reduced classifier scores
from sklearn.metrics import recall_score, precision_score

acc_score = accuracy_score(preds_red, y_test)
rec_score = recall_score(y_test, preds_red, average="micro")
prec_score = precision_score(y_test, preds_red, average="micro")

print("Accuracy score:", round(acc_score,2))
print('Recall score:', round(rec_score,2))
print('Precision score:', round(prec_score,2))

Accuracy score: 0.99
Recall score: 0.99
Precision score: 0.99


8. Compute the accuracy, precision, and recall scores for the model trained on the
reduced data set using the same test set as before. What conclusions can you make
about these models and the PCA process? How do the models compare? Discuss the
trends observed for model training time and performance for the full and reduced data
sets. Tabulate the following information in your final analysis:

                                              Random Forest:

| Item | Full Data | PCA Reduced |
|------|-----------|-------------|
|**Accuracy**|1.0|.99|
|**Precision**|1.0|.99|
|**Recall**|.1.0|.99|
|**Time**|19 µs|2.39 s|

                                             Logistic Regression:

| Item | Full Data | PCA Reduced |
|------|-----------|-------------|
|**Accuracy**|1.0|.98|
|**Precision**|1.0|.98|
|**Recall**|1.0|.98|
|**Time**|291 ms|513 ms|

- What conclusions can you make
about these models and the PCA process? How do the models compare? Discuss the
trends observed for model training time and performance for the full and reduced data
sets.

The PCA reduced Random Forest model had a slightly higher accuracy score than the reduced Logistic Regression. However the reduced Random Forest model had a shorter train time. Additionally, all 4 models appear to severly overfit the data. Oddly, the training times for the full models are shorter than the training times for the reduced-data-trained models. 